ENH: review return values for PyArray_DescrNew

mattip opened the pull request:
Fixes #19038

Review all the places `PyArray_DescrNew` and first derivative functions are used, and make sure to check the return value. Functions grepped for:
- `PyArray_DescrNew` (API function)
- `PyArray_DescrNewFromType` (API function)
- `PyArray_DescrNewByteOrder` (API function)
- `PyArray_DESCR_REPLACE` (macro)
- `ensure_dtype_nbo` (helper function)
- `_descriptor_from_pep3118_format` (helper function)
- `arraydescr_newbyteorder` (helper function)

An alternative would be to use `exit(1)` where malloc fails in `PyArray_DescrNew` since allocating a new descriptor should only fail on out-of-memory, and there is no real way to recover from that. Here are a few recommendations for using `exit(1)` (thanks @rgommers):
- https://titanwolf.org/Network/Articles/Article?AID=55f68680-740b-4f0b-ad05-a991894108bb.
  > It can be concluded that for applications above 32 bits, the "out of memory" error handler is useless

- https://eli.thegreenplace.net/2009/10/30/handling-out-of-memory-conditions-in-c
  > The abort policy is simple and familiar: when no memory is available, print a polite error message and exit (abort) the application. This is the most commonly used policy - most command-line tools and desktop applications use it.
  > As a matter of fact, this policy is so common that most Unix programs use a gnulib library function [`xmalloc`](https://manpages.ubuntu.com/manpages/xenial/man3/xmalloc.3pub.html) instead of malloc

mattip committed:
ENH: review return value from PyArray_DescrNew* calls

mattip commented on numpy/core/src/multiarray/_multiarray_tests.c.src:
```diff
@@ -649,8 +649,7 @@ fromstring_null_term_c_api(PyObject *dummy, PyObject *byte_obj)
     }
-    descr = PyArray_DescrNewFromType(NPY_FLOAT64);
-    return PyArray_FromString(string, -1, descr, -1, " ");
+    return PyArray_FromString(string, -1, NULL, -1, " ");
```
Allow PyArray_FromString to allocate the descriptor: https://github.com/numpy/numpy/blob/c65bc212ec1987caefba0ea7efe6a55803318de9/numpy/core/src/multiarray/ctors.c#L3743-L3746

mattip commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -668,6 +668,11 @@ PyArray_NewFromDescr_int(
+        PyErr_Format(PyExc_ValueError,
+                "NULL descr in call to PyArray_NewFromDescr*");
+        return NULL;
+    }
```
This is the only one of these that could actually be tested.

mattip replied:
Ahh, right, we should set the error in the API functions not here.

mattip replied:
resolved: only checking here, not setting a new error

mattip committed:
BUG: remove unused variable

mattip committed:
BUG: typo

seberg commented on numpy/core/src/multiarray/getset.c:
```diff
@@ -698,12 +698,18 @@ _get_part(PyArrayObject *self, int imag)
     offset = (imag ? type->elsize : 0);
 
     if (!PyArray_ISNBO(PyArray_DESCR(self)->byteorder)) {
         PyArray_Descr *new;
         new = PyArray_DescrNew(type);
+        if (new == NULL) {
+            return NULL;
+        }
         new->byteorder = PyArray_DESCR(self)->byteorder;
         Py_DECREF(type);
         type = new;
```
```suggestion
        Py_SETREF(type, PyArray_DescrNew(type));
        if (type == NULL) {
            return NULL;
        }
        type->byteorder = PyArray_DESCR(self)->byteorder;
```

(Currently, missing a `DECREF` in the error path, but this or the replacer macro seem easier.)

seberg commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -1337,6 +1337,9 @@ array_sort(PyArrayObject *self,
         }
         newd = PyArray_DescrNew(saved);
+        if (newd == NULL) {
+            return NULL;
```
```suggestion
            Py_DECREF(new_name);
            return NULL;
```

seberg commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -1462,6 +1465,9 @@ array_argsort(PyArrayObject *self,
         }
         newd = PyArray_DescrNew(saved);
+        if (newd == NULL) {
+            return NULL;
```
```suggestion
            Py_DECREF(new_name);
            return NULL;
```

seberg commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -1519,6 +1525,9 @@ array_argpartition(PyArrayObject *self,
         }
         newd = PyArray_DescrNew(saved);
+        if (newd == NULL) {
+            return NULL;
```
```suggestion
            Py_DECREF(new_name);
            return NULL;
```

seberg commented on numpy/core/src/multiarray/nditer_constr.c:
```diff
@@ -1132,6 +1132,9 @@ npyiter_prepare_one_operand(PyArrayObject **op,
+                    return 0;
+                }
                 Py_DECREF(*op_dtype);
                 *op_dtype = nbo_dtype;
```
Can't suggest a full change unfortunately, I would consider removing the `nbo_dtype` completely:
```
Py_SETREF(*op_dtype, PyArray_DescrNewByteorder(*op_dtype, NPY_NATIVE));
if (*op_dtype == NULL) {
    return 0;
}
```
simpler, and currently I think `*op_dtype` may be missing (depends on whether the caller expects cleanup on error).

mattip replied:
done

seberg commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -3214,6 +3214,10 @@ void_arrtype_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+        }
         ((PyVoidScalarObject *)ret)->descr->elsize = (int) memu;
         ((PyVoidScalarObject *)ret)->flags = NPY_ARRAY_BEHAVED |
                                              NPY_ARRAY_OWNDATA;
```
Annoying, but could you move the static stuff (flags + base) further up.  And then simply call `Py_DECREF(ret)` for cleanup?  The dealloc function can deal with NULL'ed `descr` just fine, but it does need the OWNDATA flag.

Right now, we are not cleaning up the object itself.

mattip replied:
done

seberg commented on numpy/core/src/umath/ufunc_type_resolution.c:
```diff
@@ -641,7 +641,7 @@ PyUFunc_SimpleUniformOperationTypeResolver(
     if (PyUFunc_ValidateCasting(ufunc, casting, operands, out_dtypes) < 0) {
         for (int iop = 0; iop < nop; iop++) {
-            Py_DECREF(out_dtypes[iop]);
+            Py_XDECREF(out_dtypes[iop]);
```
```suggestion
            Py_DECREF(out_dtypes[iop]);
```
I understand the intent, but I honestly think there is no point.  `ValidateCasting` would barf on this already if this actually happened.

mattip replied:
So maybe a comment would help future me understand.

mattip replied:
I don't understand. If one of the out_dtypes is NULL, `PyUFunc_ValidateCasting` will set an error and return -1. It would still loop through the deallocation.

seberg replied:
The `operands` can be NULL (for outputs), but `out_dtypes` cannot be?

seberg reviewed (commented):
Thanks!  Tricky stuff (although meaningless probability for running into)... a few suggestions that should be good, and 2 places where suggesting didn't do the trick.  Let me know if you prefer me applying those changes.

seberg commented:
I don't mind accepting `NULL` in for `PyArray_NewFromDescr` (passing through the error).  I generally prefer if the caller sanitizes, but it seems a good exception to me here.

mattip committed:
Update numpy/core/src/multiarray/methods.c

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

mattip committed:
Update numpy/core/src/multiarray/methods.c

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

mattip committed:
Update numpy/core/src/multiarray/getset.c

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

mattip committed:
Update numpy/core/src/multiarray/methods.c

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

mattip commented:
`PyArray_NewFromDescr` is an API function, so it is on us to check the inputs.

mattip committed:
fixes from review

mattip commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -958,6 +961,18 @@ PyArray_NewFromDescr(
+    if (descr == NULL) {
+#ifdef __GNUC__
+#pragma GCC diagnostic pop
+#endif
```
Not sure why I had to work so hard to do this check. The function is NUMPY_API, so it is on us to sanitize input. Why does gcc think descr can never be NULL?

seberg replied:
I don't agree that API automatically means we have to sanitize, I do not think the Python C-API does this.  But it _is_ fairly common for us to do, and in this case I think the pattern "relying" on it (at least theoretically) is common enough to do it.

The generated header include:
```
NPY_NO_EXPORT NPY_STEALS_REF_TO_ARG(2) NPY_GCC_NONNULL(1) NPY_GCC_NONNULL(2) PyObject * PyArray_NewFromDescr \
       (PyTypeObject *, PyArray_Descr *, int, npy_intp const *, npy_intp const *, void *, int, PyObject *);
```
So we mark this as NONNULL explicitly, actually.  This is done here:
https://github.com/numpy/numpy/blob/855b5ed82381c088a0dbf8e629927a7f337e2dbb/numpy/core/code_generators/numpy_api.py#L140

mattip replied:
Maybe we can get away with not sanitizing inputs (I think best practices is always to be liberal with input and strict with output), but using NonNull on API interfaces seems to me to be an anti-pattern. I will open an issue.

mattip replied:
Done in #20971

seberg replied:
No idea about the merit of having this in the public API.  I think we should just change it to `NonNull([1])` for the sake of this PR, though?

mattip replied:
If we do we can only put it into 1.23

seberg replied:
Hmmm, isn't that fully ABI/API compatible, since it is just an attribute (and relaxing an attribute at that)?

seberg replied:
Hmm, lets discuss tomorrow or so.  I thought we may not have to put any error at all (anywhere), because passing in NULL should only happen if the `NULL` exists due to an error being already in-progress.  And if that is not true, Python will set a `SystemError`...

mattip replied:
Let's leave that for #20971. For this PR, I pushed the check one level down.

mattip replied:
I removed the check.

mattip committed:
Update numpy/core/src/umath/ufunc_type_resolution.c

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

mattip committed:
move check to internal function

mattip committed:
remove check

seberg committed:
Remove unnecessary dealloc

The dealloc is now part of the Py_DECREF(ret) and handled there.
Doing it here would decref it twice.

seberg committed:
MAINT: Remove custom error message (and small cleanup)

It is probably not good to call PyObject_GetIter() if dtype is NULL
and an error is already in progress...
(If we check for it, lets try to do it right.)

seberg reviewed (approved):
I pushed a small fixup and removed all custom error messages (to be clear, I am OK with re-adding them).

So effectively, this fixes up a bit of shady error checks, and otherwise now semi-officially allows calling a few/most(?) array-creation functions with a `NULL` dtype.  These functions will then return `NULL` immediately while assuming that an error is already set.

seberg committed:
Fixup DescrNewFromType

`DescrNewFromType` cannot fail in most cases, but if it does,
DescrNew does not accept NULL as input.

charris commented:
Thanks Matti, thanks Sebastian.

charris commented:
I notice that numpy.sort is also part of the CVE, did we fix that?

seberg commented:
yes, it was.
